indexer/src/main/java/bio/terra/tanagra/indexing/job/bigquery/WriteEntityLevelDisplayHints.java

```diff
@@ -407,33 +418,49 @@ private List<Pair<ValueDisplay, Long>> computeEnumHintForValueDisplay(
     return enumCounts;
   }
 
-  private List<Pair<Literal, Long>> computeEnumHintForRepeatedStringValue(
-      Attribute attribute, boolean isDryRun) {
-    // TODO: Consolidate the logic here with the ValueDisplay enum hint method.
+  public static String buildEnumHintForValueDisplaySql(
+      ITEntityMain indexAttributesTable, Attribute attribute, String innerSqlSelector) {
     // Build the query.
-    // SELECT flattenedAttrVal AS enumVal, COUNT(*) AS enumCount FROM indextable
-    // CROSS JOIN UNNEST(indextable.attrVal) AS flattenedAttrVal
-    // GROUP BY enumVal
+    // SELECT attrVal AS enumVal[, attrDisp AS enumDisp], COUNT(*) AS enumCount FROM indextable
+    // GROUP BY enumVal[, enumDisp]
     SqlField attrValField = indexAttributesTable.getAttributeValueField(attribute.getName());
-    final String enumValAlias = "enumVal";
-    final String enumCountAlias = "enumCount";
-    final String flattenedAttrValAlias = "flattenedAttrVal";
-    SqlField flattenedAttrValField = SqlField.of(flattenedAttrValAlias);
+    SqlField attrDispField =
+        attribute.isValueDisplay()
+            ? indexAttributesTable.getAttributeDisplayField(attribute.getName())
+            : null;
+
+    String innerSuffix = StringUtils.EMPTY;
+    if (innerSqlSelector != null) {
+      innerSuffix = " WHERE " + innerSqlSelector;
+    }
 
     String selectEnumCountSql =
-        "SELECT "
-            + SqlQueryField.of(flattenedAttrValField, enumValAlias).renderForSelect()
-            + ", COUNT(*) AS "
-            + enumCountAlias
+        "SELECT " + SqlQueryField.of(attrValField, ENUM_VAL_ALIAS).renderForSelect();
+    if (attribute.isValueDisplay()) {
+      selectEnumCountSql +=
+          ", " + SqlQueryField.of(attrDispField, ENUM_DISP_ALIAS).renderForSelect();
+    }
+    selectEnumCountSql +=
+        ", COUNT(*) AS "
+            + ENUM_COUNT_ALIAS
             + " FROM "
             + indexAttributesTable.getTablePointer().render()
-            + " CROSS JOIN UNNEST("
-            + SqlQueryField.of(attrValField).renderForSelect()
-            + ") AS "
-            + flattenedAttrValAlias
+            + innerSuffix
             + " GROUP BY "
-            + SqlQueryField.of(attrValField, enumValAlias).renderForGroupBy(null, true);
+            + SqlQueryField.of(attrValField, ENUM_VAL_ALIAS).renderForGroupBy(null, true);
+    if (attribute.isValueDisplay()) {
+      selectEnumCountSql +=
+          ", " + SqlQueryField.of(attrDispField, ENUM_DISP_ALIAS).renderForGroupBy(null, true);
+    }
+
     LOGGER.info("SQL enum count: {}", selectEnumCountSql);
+    return selectEnumCountSql;
+  }
+
+  private List<Pair<Literal, Long>> computeEnumHintForRepeatedStringValue(
```

Comment:
I maybe don't have a complete enough understanding of the code to understand how everything fits together, but it seems like making `computeEnumHintForRepeatedStringValue` and the others public rather than `buildEnumHintForRepeatedStringValueSql` would be a cleaner interface. It seems like this creates an awkward interface where the caller is responsible for unnecessary details (e.g. reading the results). It looks like the main issue is how the SQL is actually run and perhaps that could be solved by passing in a function that's responsible for the execution but the processing could be done here regardless? Optionally there could public wrappers here (e.g. `computeEnumHintForRepeatedStringValue[Indexing|Runtime]` which supply the execution function so those don't need to be public either?

Reply:
Probably the function names add the the confusion.
The challenge here is that indexing and the user runtime (current flow) handle outputs differently. They are similar, but optimized for the required end result. I tried to avoid too many changes in the indexing flow earlier. With the filter issue fixed (the where clause added on primary entity is removed), i can add that extra logic in the current changes with minimal perf addition

Reply:
looking into it further - changes will require quite a bit of refactor in both indexer and underlay if we want to share the logic to process rows. 

- indexer code is written for big query, hence it does not use any of the Tanagra platform agnostic data models. Instead it uses BQ native data types. example: tableResult is a BQ construct that is converted to SqlQueryResult. 
- underlay only exposes platform independent SqlQueryResult (and other interfaces), though internally it calls  the same BQ functions. Additionally even the executor is platform specific, so it can't be exposed in the underlay interface.
- indexer cannot added as a dependency to underlay (creates a circular dependency) 

so, I would prefer to keep it as it is in the PR - open to renaming the functions if preferred. 
I will however check if i can extract the processing of a single row to a static util

Reply:
update - even the individual rows do not align in data types, its best to avoid major refactor now.

Reply:
Okay, sounds good.